runners/samza/src/main/java/org/apache/beam/runners/samza/renderer/PipelineJsonRenderer.java

```diff
@@ -0,0 +1,196 @@
+    visitor.begin();
+    pipeline.traverseTopologically(visitor);
+    visitor.end();
+    return visitor.jsonBuilder.toString();
+  }
+
+  public static String toJsonString(RunnerApi.Pipeline pipeline) {
+    return null;
+  }
+
+  private final StringBuilder jsonBuilder = new StringBuilder();
+  private final StringBuilder graphLinks = new StringBuilder();
+  private final Map<TransformHierarchy.Node, Integer> nodeToId = new HashMap<>();
+  private final Map<PValue, String> valueToProducerNodeName = new HashMap<>();
+  private int indent;
+  private int nextNodeId;
+
+  private PipelineJsonRenderer() {}
+
+  @Override
+  public void enterPipeline(Pipeline p) {}
+
+  @Override
+  public CompositeBehavior enterCompositeTransform(TransformHierarchy.Node node) {
+    String fullName = node.getFullName();
+    writeLine(
+        "{ \"fullName\":\"%s\",", escapeString(fullName.isEmpty() ? OUTERMOST_NODE : fullName));
+    writeLine(
+        "  \"shortName\":\"%s\",",
+        escapeString(fullName.isEmpty() ? OUTERMOST_NODE : node.getTransform().getName()));
+    writeLine("  \"id\":\"%s\",", escapeString(fullName.isEmpty() ? OUTERMOST_NODE : fullName));
+    if (!fullName.isEmpty()) {
+      String enclosingNodeName = node.getEnclosingNode().getFullName();
+      writeLine(
+          "  \"enclosingNode\":\"%s\",",
+          escapeString(enclosingNodeName.isEmpty() ? OUTERMOST_NODE : enclosingNodeName));
+    }
+
+    String ioInfo = getIOTopicInfo(node);
+    if (!ioInfo.isEmpty()) {
+      writeLine(" \"ioInfo\":\"%s\",", escapeString(ioInfo));
+    }
+
+    writeLine("  \"ChildNode\":[");
+    enterBlock();
+    return CompositeBehavior.ENTER_TRANSFORM;
+  }
+
+  @Override
+  public void leaveCompositeTransform(TransformHierarchy.Node node) {
+    exitBlock();
+    writeLine("]},");
+  }
+
+  @Override
+  public void visitPrimitiveTransform(TransformHierarchy.Node node) {
+    String fullName = node.getFullName();
+    writeLine("{ \"fullName\":\"%s\",", escapeString(fullName));
+    writeLine("  \"shortName\":\"%s\",", escapeString(node.getTransform().getName()));
+    String enclosingNodeName = node.getEnclosingNode().getFullName();
+    writeLine(
+        "  \"enclosingNode\":\"%s\",",
+        escapeString(enclosingNodeName.isEmpty() ? OUTERMOST_NODE : enclosingNodeName));
+    writeLine("  \"id\":\"%s\"},", escapeString(fullName));
+
+    node.getOutputs().values().forEach(x -> valueToProducerNodeName.put(x, fullName));
+    node.getInputs()
+        .forEach(
+            (key, value) -> {
+              final String producerName = valueToProducerNodeName.get(value);
+              String style = "solid";
+              if (node.getTransform().getAdditionalInputs().containsKey(key)) {
+                style = "dashed";
+              }
+              graphLinks.append(
+                  String.format(
+                      "{\"from\":\"%s\"," + "\"to\":\"%s\"," + "\"hashId\":\"%s\"},",
```

Comment:
what is the use of hash id in the json graph?

Reply:
It is generated by the TupleTag which has a unique ID.

Reply:
I think my question is, why do we want to include the unique id in the json graph? What is it used for ?

Reply:
Actually we don't need this unique ID since I'm not using it further ahead. Hence, I have removed this. Thanks for pointing this out.